// discord-bot-quickstart/lib/console-reader/console-reader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConsoleReader = void 0;
const helpers_1 = require("../helpers");
const readline = require("readline");
const minimist = require("minimist");
class ConsoleReader {
    constructor(logger) {
        this.logger = logger;
        this.commands = new helpers_1.CommandMap();
    }
    listen() {
        let rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        rl.on('line', input => {
            if (!input)
                return;
            let parts = input.split(' ');
            let result = minimist(parts);
            if (this.commands.has(result._[0])) {
                let cmds = this.commands.get(result._[0]);
                cmds.forEach(cmd => cmd(result, rl));
            }
        });
        rl.on('close', () => {
            this.logger.debug('Console Reader Disconnected');
        });
    }
}
exports.ConsoleReader = ConsoleReader;
